cli/flox/src/commands/init/node.rs

```diff
@@ -69,38 +98,153 @@ enum NodeAction {
+        let nvmrc_version = if package_manager.is_some() {
             None
+        } else {
+            match package_json_node_version {
+                // package.json is higher priority than .nvmrc,
+                // so don't check .nvmrc if we know we'll use the version in
+                // package.json or we know we can't provide it
+                PackageJSONVersion::Found(_) | PackageJSONVersion::Unavailable => None,
+                _ => Self::get_nvmrc_version(path, flox)?,
+            }
         };
 
         Ok(Self {
+            package_json_node_version,
             nvmrc_version,
-            package_json_version,
-            hook,
+            package_manager,
         })
     }
 
+    /// Look for nodejs, npm, and yarn versions in a (possibly non-existent)
+    /// `package.json` file
+    fn get_package_json_versions(path: &Path) -> Result<Option<PackageJSONVersions>> {
+        let package_json = path.join("package.json");
+        if !package_json.exists() {
+            return Ok(None);
+        }
+        let package_json_contents = fs::read_to_string(package_json)?;
+        match serde_json::from_str::<serde_json::Value>(&package_json_contents) {
+            // Treat a package.json that can't be parsed as JSON the same as it not existing
+            Err(_) => Ok(None),
+            Ok(package_json_json) => {
+                let node = package_json_json["engines"]["node"]
+                    .as_str()
+                    .map(|s| s.to_string());
+                let npm = package_json_json["engines"]["npm"]
+                    .as_str()
+                    .map(|s| s.to_string());
+                let yarn = package_json_json["engines"]["yarn"]
+                    .as_str()
+                    .map(|s| s.to_string());
+                Ok(Some(PackageJSONVersions { node, npm, yarn }))
+            },
+        }
+    }
+
+    /// Try to find node, npm, and yarn versions that satisfy constraints in
+    /// package.json
+    fn get_package_manager(
+        versions: &PackageJSONVersions,
+        package_lock_exists: bool,
+        yarn_lock_exists: bool,
+        flox: &Flox,
+    ) -> Result<Option<PackageManager>> {
+        let PackageJSONVersions { npm, yarn, node } = versions;
+
+        let found_node = match node {
+            Some(node_version) => {
+                match Self::get_default_node_if_compatible(Some(node_version.clone()), flox)? {
+                    // If the corresponding node isn't compatible, don't install a package manager
+                    None => return Ok(None),
+                    Some(found_node) => found_node,
+                }
+            },
+            None => Self::get_default_node_if_compatible(None, flox)?
+                .ok_or(anyhow!("Flox couldn't find nodejs in nixpkgs"))?,
+        };
+
+        // We assume that yarn and npm are built with found_node, which is
+        // currently true in nixpkgs
```

Comment:
We use `get_default_node_if_compatible` to get `found_node` which sets `rel_path` to `nodejs`. I agree that's not bulletproof, but I think it should hopefully age okay?

Reply:
It's no longer an issue for npm, because we're never installing npm.

For yarn, we're saying that nixpkgs#yarn is built with nixpkgs#nodejs. Based on https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/all-packages.nix#L15014 and https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/yarn/default.nix#L1, I don't expect that will change, but you're right that it could. Even if it does change, we won't have two versions in the closure, because we're installing `yarn` but not `nodejs`.